src/rust/ide/lib/view/src/graph_editor/component/project_name.rs

```diff
@@ -0,0 +1,327 @@
+use ensogl::display::scene::Scene;
+use ensogl::display::shape::text::text_field::TextField;
+use ensogl::display::shape::text::text_field::TextFieldProperties;
+use ensogl::display::shape::*;
+use ensogl::display::Sprite;
+use ensogl::display::world::World;
+use ensogl::gui::component::Animation;
+use ensogl::gui::component;
+use logger::enabled::Logger;
+use logger::AnyLogger;
+use nalgebra::Vector2;
+use ensogl::animation::linear_interpolation;
+
+
+
+// =================
+// === Constants ===
+// =================
+
+const TEXT_SIZE              : f32         = 16.0;
+const TEXT_COLOR             : color::Rgba = color::Rgba::new(1.0, 1.0, 1.0, 1.0);
+const TRANSPARENT_TEXT_COLOR : color::Rgba = color::Rgba::new(1.0, 1.0, 1.0, 0.6);
+
+/// Project name used as a placeholder in ProjectName view when it's initialized.
+pub const UNKNOWN_PROJECT_NAME:&str = "[UNKNOWN NAME]";
+
+
+
+// ==================
+// === Background ===
+// ==================
+
+mod background {
+    use super::*;
+
+    ensogl::define_shape_system! {
+            (style:Style, selection:f32) {
+                let bg_color = color::Rgba::new(0.0,0.0,0.0,0.000_001);
+                Plane().fill(bg_color).into()
+            }
+        }
+}
+
+
+
+// =================
+// === FrpInputs ===
+// =================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpInputs {
+    /// Rename the project.
+    pub name : frp::Source<String>,
+    /// Reset the project name to the one before editing.
+    pub cancel_editing : frp::Source
+}
+
+impl FrpInputs {
+    /// Create new FrpInputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def cancel_editing = source();
+            def name           = source();
+        }
+        Self{cancel_editing,name}
+    }
+}
+
+
+
+// ==================
+// === FrpOutputs ===
+// ==================
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct FrpOutputs {
+    /// Emits the new project name when it's renamed.
+    pub name : frp::Source<String>
+}
+
+impl FrpOutputs {
+    /// Create new FrpOutputs.
+    pub fn new(network:&frp::Network) -> Self {
+        frp::extend! {network
+            def name = source();
+        }
+        Self{name}
+    }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Frp {
+    pub inputs  : FrpInputs,
+    pub outputs : FrpOutputs,
+    pub network : frp::Network,
+}
+
+impl Deref for Frp {
+    type Target = FrpInputs;
+    fn deref(&self) -> &Self::Target {
+        &self.inputs
+    }
+}
+
+impl Default for Frp {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Frp {
+    /// Create new Frp.
+    pub fn new() -> Self {
+        let network = frp::Network::new();
+        let inputs  = FrpInputs::new(&network);
+        let outputs = FrpOutputs::new(&network);
+        Self{network,inputs,outputs}
+    }
+}
```

Comment:
Ok, this needs to be written in guide, but I would bethankful if we will be using the following layout:
- struct definition
- explicit Deref impl if any
- methods definitions
- all other impls

The reason is that Deref is important while inspectingthe layout of structure, so youshould know the structimplements it immediately. Methods are your public API(with deref), while all other impls impleent additional utilsand there could be a ton of impls andyoudontwant to have methods 500 lines below struct

Reply:
@wdanilo sorry, maybe I am too sleepy. But what's the expected action here?

Reply:
Relayouting the code according ot the poitns there - write to me on Discord if that's still not clear :)

Reply:
Oh, done. I need to relayout https://github.com/enso-org/ide/pull/662 to follow this guideline.